src/server/generic_family.cc

```diff
@@ -243,161 +251,182 @@ OpResult<RestoreArgs> RestoreArgs::TryFrom(const CmdArgList& args) {
 
 OpStatus OpPersist(const OpArgs& op_args, string_view key);
 
+OpResult<std::string> OpDump(const OpArgs& op_args, string_view key);
+
+OpResult<bool> OnRestore(const OpArgs& op_args, std::string_view key, std::string_view payload,
+                         RestoreArgs restore_args, int rdb_version);
+
 class Renamer {
  public:
-  Renamer(ShardId source_id) : src_sid_(source_id) {
+  Renamer(std::string_view src_key, std::string_view dest_key, unsigned shard_count)
+      : src_key_(src_key),
+        dest_key_(dest_key),
+        src_sid_(Shard(src_key, shard_count)),
+        dest_sid_(Shard(dest_key, shard_count)) {
   }
 
-  void Find(Transaction* t);
+  void Initialize(Transaction* t);
 
   OpResult<void> status() const {
     return status_;
   };
 
-  void Finalize(Transaction* t, bool skip_exist_dest);
+  void Finalize(Transaction* t, bool destination_should_not_exist);
 
  private:
-  OpStatus MoveSrc(Transaction* t, EngineShard* es);
-  OpStatus UpdateDest(Transaction* t, EngineShard* es);
+  bool KeyExists(Transaction* t, EngineShard* shard, std::string_view key);
+  OpStatus SerializeSrc(Transaction* t, EngineShard* shard);
 
-  ShardId src_sid_;
+  OpStatus DelSrc(Transaction* t, EngineShard* shard);
+  OpStatus DeserializeDest(Transaction* t, EngineShard* shard);
 
-  struct FindResult {
-    string_view key;
-    PrimeValue ref_val;
-    uint64_t expire_ts;
+  struct SerializedValue {
+    std::string value;
+    int64_t expire_ts;
     bool sticky;
-    bool found = false;
   };
 
-  PrimeValue pv_;
-  string str_val_;
+ private:
+  const std::string_view src_key_;
+  const std::string_view dest_key_;
+  const ShardId src_sid_;
+  const ShardId dest_sid_;
+
+  bool src_found_ = false;
+  bool dest_found_ = false;
+
+  SerializedValue serialized_value_;
 
-  FindResult src_res_, dest_res_;  // index 0 for source, 1 for destination
   OpResult<void> status_;
 };
 
-void Renamer::Find(Transaction* t) {
+void Renamer::Initialize(Transaction* t) {
   auto cb = [this](Transaction* t, EngineShard* shard) {
     auto args = t->GetShardArgs(shard->shard_id());
     DCHECK_EQ(1u, args.Size());
 
-    FindResult* res = (shard->shard_id() == src_sid_) ? &src_res_ : &dest_res_;
+    const ShardId shard_id = shard->shard_id();
 
-    res->key = args.Front();
-    auto& db_slice = EngineShard::tlocal()->db_slice();
-    auto [it, exp_it] = db_slice.FindReadOnly(t->GetDbContext(), res->key);
+    if (shard_id == src_sid_) {
+      src_found_ = KeyExists(t, shard, src_key_);
```

Comment:
OpDump also checks if key exists and if not it returns KEY_NOTFOUND, so we can instead of calling KeyExists here and do find twice on dbslice just call SerializeSrc and check the return status